Sources/FireKit/UserSessionManager.swift

```diff
@@ -9,224 +9,232 @@ import Foundation
-  }
-  
-  // MARK: - Password Changes
-  
-  open func changePasswordRequest(replace oldPassword: String, with newPassword: String, completion: ((Result<Void, Error>) -> Void)? = nil) {
-    let user = Auth.auth().currentUser
-    let credentials = EmailAuthProvider.credential(withEmail: email, password: oldPassword)
-    user?.reauthenticate(with: credentials, completion: { (result, error) in
-      if let error = error {
-        completion?(.failure(error))
-      } else {
-        // User re-authenticated.
-        user?.updatePassword(to: newPassword){ error in
-          if let error = error { completion?(.failure(error)) }
-          else { completion?(.success(())) }
+    
+    @Published public var userReady: Bool = false
+    @Published public var isNew = false
+    @Published public var state: PTAuthenticationState = .initializing
+    
+    @Published public var userID: String? {
+        didSet { userReady = getUserReady() }
+    }
+    
+    let defaultToAnonymousUser: Bool
+    let onAuth: ((User) -> Void)?
+    
+    public init(defaultToAnonymousUser: Bool = false, onAuth: ((User) -> Void)? = nil ){
+        self.defaultToAnonymousUser = defaultToAnonymousUser
+        self.onAuth = onAuth
+    }
+    
+    public var initialized: Bool {
+        state != .initializing
+    }
+    
+    public var isAnonymous: Bool {
+        Auth.auth().currentUser?.isAnonymous ?? false
+    }
+    
+    public var email: String {
+        Auth.auth().currentUser?.email ?? ""
+    }
+    
+    
+    // MARK: - Exposed Callbacks
+    
+    open func onAnonymousSessionStart() { }
+    open func getUserReady() -> Bool {
+        userID != nil
+    }
+    
+    // Called on new FirebaseAuth sessions, triggered by Auth.stateChangedListener started in startAuthListener
+    open func onAuthSessionStart(user: FirebaseAuth.User?, error: Error?) {
+        print("PTUserSessionManager: Auth Session Started")
+        guard let user = user else {
+            print("PTUserSessionManager: Auth Session Started (No User)")
+            state = .unauthenticated
+            userID = nil
+            return
+        }
+        
+        userID = user.uid
+        state = .authenticated
+        onAuth?(user)
+    }
+    
+    open func onAuthSessionEnd() {
+        print("PTUserSessionManager: Auth Session Ended")
+    }
+    
+    // MARK: - Auth Session Lifecycle
+    
+    open func startAnonymousSession() {
+        print("PTUserSessionManager: Starting Anonymous Session")
+        Auth.auth().signInAnonymously() { (authResult, error) in
+            if let error = error {
+                print("PTUserSessionManager: Starting Anonymous Session (ERROR)")
+                print("PTUserSessionManager: ", error)
+                self.state = .unauthenticated
+                return
+            }
+            print("PTUserSessionManager: Starting Anonymous Session (STARTED)")
+            self.state = .authenticated
+            self.isNew = true
```

Comment:
bu isNew logic'i çalışmıyor işte, when this action is completed, `onAuthStateChanged` çağrılıyor ilk, çünkü completion'ın çalışması için state'in değişmesi lazım. Ondan önce user logged in state'e giriyor, sonra isNew'a geri jump ediyor gibi garip bir şey var. Routing'i tabi swiftui'siz ben stateful yapamadığımdan bu problem daha emphasized, enum'a bağlı yapsam less noticable olacak.

Reply:
ben bunu anlamadim. `isNew` `onAuthStateChanged`'de degismiyor ki? signupta true oluyor. Sonra bir daha olmuyor. Burda race condition sadece signup tamamen bitmeden once onu kullanmaya calisirsan oluyor. Baska bir hatalar da vardi onlari bu PR cozuyor

Reply:
e tamam ben de diyorum o race condition benim için gerçekleşiyor çünkü onAuthStateChanged'de isNew'sa bir şey yaptırmak istiyorum, fakat isnew false geliyor

Reply:
specifically onAuthSessionStart olunca bir şey yaptırmak istiyorum, eğer user new'sa

Reply:
anladim evet. bunu cozmustuk signupta isNew'u disari alarak. anon icin de self.isNew = true satirini line 82ye alcaz